File: kvx_flutter/lib/viewmodels/device_viewmodel.dart
import 'package:flutter/foundation.dart';
import '../models/device.dart';

enum DeviceFilter { all, online, offline, busy }

extension DeviceFilterExtension on DeviceFilter {
  String get displayName {
    switch (this) {
      case DeviceFilter.all:
        return 'All';
      case DeviceFilter.online:
        return 'Online';
      case DeviceFilter.offline:
        return 'Offline';
      case DeviceFilter.busy:
        return 'Busy';
    }
  }
}

class DeviceViewModel extends ChangeNotifier {
  List<Device> _devices = Device.sampleDevices;
  bool _isLoading = false;
  String _searchText = '';
  DeviceFilter _selectedFilter = DeviceFilter.all;

  List<Device> get devices => _devices;
  bool get isLoading => _isLoading;
  String get searchText => _searchText;
  DeviceFilter get selectedFilter => _selectedFilter;

  List<Device> get filteredDevices {
    var result = _devices;

    if (_searchText.isNotEmpty) {
      result = result
          .where((d) => d.name.toLowerCase().contains(_searchText.toLowerCase()))
          .toList();
    }

    switch (_selectedFilter) {
      case DeviceFilter.all:
        break;
      case DeviceFilter.online:
        result = result.where((d) => d.status == DeviceStatus.online).toList();
        break;
      case DeviceFilter.offline:
        result = result.where((d) => d.status == DeviceStatus.offline).toList();
        break;
      case DeviceFilter.busy:
        result = result.where((d) => d.status == DeviceStatus.busy).toList();
        break;
    }

    return result;
  }

  int get onlineCount => _devices.where((d) => d.status == DeviceStatus.online).length;
  int get offlineCount => _devices.where((d) => d.status == DeviceStatus.offline).length;
  int get busyCount => _devices.where((d) => d.status == DeviceStatus.busy).length;

  void setSearchText(String value) {
    _searchText = value;
    notifyListeners();
  }

  void setFilter(DeviceFilter filter) {
    _selectedFilter = filter;
    notifyListeners();
  }

  Future<void> refresh() async {
    _isLoading = true;
    notifyListeners();

    await Future.delayed(const Duration(seconds: 1));

    _isLoading = false;
    notifyListeners();
  }

  void deleteDevices(List<Device> devicesToDelete) {
    for (final device in devicesToDelete) {
      _devices.removeWhere((d) => d.id == device.id);
    }
    notifyListeners();
  }

  void addDevice({required String name, required DeviceType type, required DeviceStatus status}) {
    _devices.add(Device(name: name, type: type, status: status));
    notifyListeners();
  }

  void toggleStatus(Device device) {
    final index = _devices.indexWhere((d) => d.id == device.id);
    if (index != -1) {
      final currentStatus = _devices[index].status;
      DeviceStatus newStatus;

      switch (currentStatus) {
        case DeviceStatus.online:
          newStatus = DeviceStatus.offline;
          break;
        case DeviceStatus.offline:
          newStatus = DeviceStatus.online;
          break;
        case DeviceStatus.busy:
          newStatus = DeviceStatus.online;
          break;
      }

      _devices[index] = device.copyWith(status: newStatus);
      notifyListeners();
    }
  }
}

File: kvx/ViewModels/DeviceViewModel.swift
//
//  DeviceViewModel.swift
//  kvx
//
//  Created by Vinh Nguyen on 22/4/26.
//

import Foundation
import SwiftUI

@Observable
final class DeviceViewModel {
    private let fetchDevicesUseCase: FetchDevicesUseCase
    var devices: [Device] = Device.sampleDevices
    var isLoading: Bool = false
    var errorMessage: String?
    var searchText: String = ""
    var selectedFilter: DeviceFilter = .all

    init(repository: DeviceRepository) {
        fetchDevicesUseCase = FetchDevicesUseCase(repository: repository)
    }

    enum DeviceFilter: String, CaseIterable {
        case all = "All"
        case online = "Online"
        case offline = "Offline"
        case busy = "Busy"
    }

    var filteredDevices: [Device] {
        var result = devices

        if !searchText.isEmpty {
            result = result.filter { $0.name.localizedCaseInsensitiveContains(searchText) }
        }

        switch selectedFilter {
        case .all:
            break
        case .online:
            result = result.filter { $0.status == .online }
        case .offline:
            result = result.filter { $0.status == .offline }
        case .busy:
            result = result.filter { $0.status == .busy }
        }

        return result
    }

    var onlineCount: Int {
        devices.filter { $0.status == .online }.count
    }

    var offlineCount: Int {
        devices.filter { $0.status == .offline }.count
    }

    var busyCount: Int {
        devices.filter { $0.status == .busy }.count
    }

    func loadDevices() async {
        isLoading = true
        errorMessage = nil
        do {
            devices = try await fetchDevicesUseCase.execute()
        } catch {
            errorMessage = error.localizedDescription
        }
        isLoading = false
    }

    func refresh() async {
        await loadDevices()
    }

    func deleteDevices(at offsets: IndexSet) {
        let devicesToDelete = offsets.map { filteredDevices[$0] }
        for device in devicesToDelete {
            if let index = devices.firstIndex(where: { $0.id == device.id }) {
                devices.remove(at: index)
            }
        }
    }

    func moveDevices(from source: IndexSet, to destination: Int) {
        let devicesToMove = source.map { filteredDevices[$0] }
        var insertIndex = destination

        for device in devicesToMove {
            if let sourceIndex = devices.firstIndex(where: { $0.id == device.id }) {
                let adjustedDestination = sourceIndex < destination ? insertIndex - 1 : insertIndex
                devices.move(fromOffsets: IndexSet(integer: sourceIndex), toOffset: adjustedDestination)
                if sourceIndex < destination {
                    insertIndex -= 1
                }
            }
        }
    }

    func addDevice(name: String, type: Device.DeviceType, status: Device.DeviceStatus, temperature: Double? = nil, humidity: Double? = nil) {
        let newDevice = Device(name: name, type: type, status: status, temperature: temperature, humidity: humidity)
        devices.append(newDevice)
    }

    func toggleStatus(for device: Device) {
        if let index = devices.firstIndex(where: { $0.id == device.id }) {
            let currentStatus = devices[index].status
            let newStatus: Device.DeviceStatus
            switch currentStatus {
            case .online:
                newStatus = .offline
            case .offline:
                newStatus = .online
            case .busy:
                newStatus = .online
            }
            devices[index] = Device(
                id: device.id,
                name: device.name,
                type: device.type,
                status: newStatus,
                temperature: device.temperature,
                humidity: device.humidity
            )
        }
    }
}
